### DISCUSSION TITLE
Collection expression for multi-dimensional arrays

### DISCUSSION THREAD

**Discussion by huoyaoyuan** [dotnet/csharplang#7871] [General]
Just noticed that MD array is supported by traditional collection initializer:

```csharp
int[,] = array = { { 1, 2 }, { 3, 4 } };
```

But not collection expression:
```csharp
int[,] array = [[1, 2], [3, 4]];
```

The potential benefit of collection expression is allowing spreads. However, MD array requires the length are the same for each dimensions. Length difference need to be handled at runtime, if dynamic length with spreads is introduced.

Has this been discussed before?

**Comment by sharpchen** [comment]
I believe the original proposal of collection expression did mention this, but said this is not possible. This syntax was already assigned for jagged array.

**Reply by sharpchen** [reply]
Current proposal was truncated, you can find **Unsupported scenarios** in #5354 edited on Jan 30, 2023

> Multi-dimensional arrays (e.g. new int[5, 10] { ... }). There is no facility to include the dimensions, and all collection literals are either linear or map structures only.

**Reply by jnm2** [reply]
This can also be seen in the C# 12 spec proposal, which is where the spec proposal went when it was moved out of #5354:
https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#unsupported-scenarios

**Comment by jnm2** [comment]
There hasn't been motivation thus far for multidimensional collection expressions. When we get natural types, I'd expect `var x = [[1, 2], [3, 4]];` to create a collection of collections. It could switch to meaning multidimensional initialization if the target type is a multidimensional array, but this does seem like a compromise of intuition. It _looks_ like a collection of collections. (To be fair, so does multidimensional initializer syntax.)

There's a lack of obvious wins that have been identified here. You mentioned spreads:

What code should the compiler generate to spread something into a multidimensional array, and how can this be at least as efficient as existing methods of initializing multidimensional arrays?

What should happen when things of unequal length are spread?

**Reply by colejohnson66** [reply]
IMO, `var x = [[1, 2], [3, 4]];` would indeed have a target type of collection-of-collections (such as `int[][]` or `List<List<int>>`. But if the target type is *known* to be `T[,]`, I would expect the language to treat it the same as the multidimensional initializer. I don't know how hard such a change would be in Roslyn, but it would probably be just a tweak in the language spec.
